src/lib/operators/index_scan.cpp

```diff
@@ -34,175 +34,111 @@ const std::string& IndexScan::name() const {
 
-std::shared_ptr<AbstractTask> IndexScan::_create_job(const ChunkID chunk_id, std::mutex& output_mutex) {
-  auto job_task = std::make_shared<JobTask>([this, chunk_id, &output_mutex]() {
-    // The output chunk is allocated on the same NUMA node as the input chunk.
-    const auto chunk = _in_table->get_chunk(chunk_id);
-    if (!chunk) {
-      return;
+      if (mapped_chunk_id != chunk_id_mapping.end()) {
+        matches_out->emplace_back(mapped_chunk_id->second, (*current_iter).chunk_offset);
+      }
     }
+  };
 
-    const auto matches_out = std::make_shared<RowIDPosList>(_scan_chunk(chunk_id));
-    if (matches_out->empty()) {
-      return;
+  switch (_predicate_condition) {
+    case PredicateCondition::Equals: {
+      index->range_equals_with_iterators(append_matches, _scan_value);
+      break;
     }
-
-    auto segments = Segments{};
-    const auto column_count = _in_table->column_count();
-    for (auto column_id = ColumnID{0}; column_id < column_count; ++column_id) {
-      auto ref_segment_out = std::make_shared<ReferenceSegment>(_in_table, column_id, matches_out);
-      segments.push_back(ref_segment_out);
+    case PredicateCondition::NotEquals: {
+      index->range_not_equals_with_iterators(append_matches, _scan_value);
+      break;
     }
-
-    const auto lock = std::lock_guard<std::mutex>{output_mutex};
-    _out_table->append_chunk(segments, nullptr, chunk->get_allocator());
-  });
-
-  return job_task;
-}
-
-void IndexScan::_validate_input() {
-  Assert(_predicate_condition != PredicateCondition::Like, "Predicate condition not supported by index scan.");
-  Assert(_predicate_condition != PredicateCondition::NotLike, "Predicate condition not supported by index scan.");
-
-  Assert(_left_column_ids.size() == _right_values.size(),
-         "Count mismatch: left column IDs and right values don’t have same size.");
-  if (is_between_predicate_condition(_predicate_condition)) {
-    Assert(_left_column_ids.size() == _right_values2.size(),
-           "Count mismatch: left column IDs and right values don’t have same size.");
+    default:
+      Fail("Unsupported comparison type, PartialHashIndex only supports Equals and NotEquals.");
   }
 
-  Assert(_in_table->type() == TableType::Data, "IndexScan only supports persistent tables right now.");
-}
+  if (matches_out->empty()) {
+    return _out_table;
+  }
 
-RowIDPosList IndexScan::_scan_chunk(const ChunkID chunk_id) {
-  const auto to_row_id = [chunk_id](ChunkOffset chunk_offset) { return RowID{chunk_id, chunk_offset}; };
+  // Split the output RowIDPosList per Chunk.
 
-  auto range_begin = AbstractChunkIndex::Iterator{};
-  auto range_end = AbstractChunkIndex::Iterator{};
+  auto matches_out_per_chunk = std::vector<std::shared_ptr<RowIDPosList>>{};
+  auto first_pos_list = std::make_shared<RowIDPosList>();
+  first_pos_list->guarantee_single_chunk();
+  matches_out_per_chunk.emplace_back(first_pos_list);
```

Comment:
I totally missed this; it would be really bad.
I don't think the matches returned by the Index are sorted Chunk-wise.
I just added this Assert to the IndexScan, and surprisingly, the TPC-C and the TPC-DS are not crashing, tho.

Reply:
I thought since insertion happens chunk-wise (if I remember the interface correctly) and we simply iterate the chunk, this order is guaranteed.
I would be happy with this assert (if you agree that we end up sorted). Otherwise, we should not create new chunks whenever the `chunk_id` changes.

Reply:
There is a case when it is not possible: NotEquals predicates.

Reply:
How much do we benefit from the single chunk guarantee? If it's not much, we can also consider to just iterate the matches and fill chunks subsequently until they reach a target (default) size.

Reply:
Currently, I tend to keep the current approach as it yields the same result as a table scan (whose result might be union'ed afterwards). Makes reasoning about the following operators somewhat easier, I hope.

For NonEqual predicates, though, I would go for the "default-chunk-size-sized PosLists".

Any opinions on that? @tjjordan @dey4ss

Reply:
Hmm, now looking at the code I tend to keep it simpler and go for the "default-chunk-size-sized PosLists" in both cases.

Can somebody throw a coin here?